konnectivity-client/pkg/client/client.go

```diff
@@ -185,14 +189,41 @@ func CreateSingleUseGrpcTunnelWithContext(createCtx, tunnelCtx context.Context,
 }
 
 func newUnstartedTunnel(stream client.ProxyService_ProxyClient, c clientConn) *grpcTunnel {
-	return &grpcTunnel{
+	t := grpcTunnel{
 		stream:             stream,
 		clientConn:         c,
 		pendingDial:        pendingDialManager{pendingDials: make(map[int64]pendingDial)},
 		conns:              connectionManager{conns: make(map[int64]*conn)},
 		readTimeoutSeconds: 10,
 		done:               make(chan struct{}),
+		prevStatus:         metrics.TunnelStatusCreated,
 	}
+	t.initMetric(metrics.TunnelStatusCreated)
+	return &t
+}
+
+// initMetric must be called exactly once to initialize tunnels metric.
+func (t *grpcTunnel) initMetric(status metrics.TunnelStatus) {
+	t.statusLock.Lock()
+	defer t.statusLock.Unlock()
+	metrics.Metrics.GetTunnelsMetric().WithLabelValues(string(status)).Inc()
+	t.prevStatus = status
+}
+
+func (t *grpcTunnel) updateMetric(status metrics.TunnelStatus) {
+	t.statusLock.Lock()
+	defer t.statusLock.Unlock()
+	m := metrics.Metrics.GetTunnelsMetric()
+	m.WithLabelValues(string(t.prevStatus)).Dec()
+	m.WithLabelValues(string(status)).Inc()
+	t.prevStatus = status
+}
+
+// closeMetric must be called exactly once to finalize tunnels metric.
```

Comment:
Yousay it should only be called once but we do nothing to ensure this is true. It also seems like the more important thing is that update should not be called after this has been called. However we do nothing to ensure that is true?

Reply:
Added some guards using the done channel.